### PR TITLE
Fix move to namespace refactoring

### LINKED ISSUE #39234: MoveToNamespaceCodeActionProvider crashed with ArgumentOutOfRangeException in multi-targetFrameworks projects
**Version Used**: 
VS2019 16.4.0Preview1

**Steps to Reproduce**:

1. Create an csproj and modify to multi targetFrameworks;
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFrameworks>netstandard2.0;net472</TargetFrameworks>
  </PropertyGroup>

</Project>

```
2. Create an cs file with 2 classes;
```cs
using System;

namespace ClassLibrary1.A
{
    public class Class1
    {
    }

    public class Class2
    {
    }
}
```
3. call `Quick Actions/Move to Namespace` on the last class `Class2`, type any namespace and click `OK`;

**Expected Behavior**:
It should work as it does with single-targetFramework projects

**Actual Behavior**:
```cs
System.ArgumentOutOfRangeException : Specified argument was out of the range of valid values.
Parameter name: span
   at Microsoft.CodeAnalysis.SyntaxNode.FindNode(TextSpan span,Boolean findInsideTrivia,Boolean getInnermostNodeForTie)
   at async Microsoft.CodeAnalysis.CSharp.ChangeNamespace.CSharpChangeNamespaceService.TryGetApplicableContainerFromSpanAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.ChangeNamespace.AbstractChangeNamespaceService`3.TryGetApplicableContainersFromAllDocumentsAsync[TNamespaceDeclarationSyntax,TCompilationUnitSyntax,TMemberDeclarationSyntax](<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CSharp.ChangeNamespace.CSharpChangeNamespaceService.GetValidContainersFromAllLinkedDocumentsAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.ChangeNamespace.AbstractChangeNamespaceService`3.ChangeNamespaceAsync[TNamespaceDeclarationSyntax,TCompilationUnitSyntax,TMemberDeclarationSyntax](<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.MoveToNamespace.AbstractMoveToNamespaceService`2.MoveItemsInNamespaceAsync[TNamespaceDeclarationSyntax,TNamedTypeDeclarationSyntax](<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.MoveToNamespace.AbstractMoveToNamespaceService`2.MoveTypeToNamespaceAsync[TNamespaceDeclarationSyntax,TNamedTypeDeclarationSyntax](<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.MoveToNamespace.AbstractMoveToNamespaceCodeAction.ComputeOperationsAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeActions.CodeActionWithOptions.GetOperationsAsync(<Unknown Parameters>)
   at Roslyn.Utilities.TaskExtensions.WaitAndGetResult_CanCallOnBackground[T](Task`1 task,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedAction.InvokeWorker(Func`1 getFromDocument,IProgressTracker progressTracker,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedAction.<>c__DisplayClass20_0.<InvokeCore>b__0()
   at Microsoft.CodeAnalysis.Extensions.IExtensionManagerExtensions.PerformAction(IExtensionManager extensionManager,Object extension,Action action)
```

### PR DESCRIPTION
**Pull request by genlu**
Fix #39234.

The cause of the crash is MoveTypeService not handling linked documents properly, so in MTFM case, we would end up with multiple linked documents with different content, which is unexpected by ChangeNamespaceService. 

It is fixed by:
1. Merge linked file diffs before sending request to ChangeNamespaceService
2. Also check for out of range span in ChangeNamespaceService

### CODE REVIEW

**Comment by genlu** [comment]
FYI @dpoeschl for the merge-diff stuff

**Comment by genlu** [comment]
@ryzngard Could you please take a look? Thanks

File: src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractChangeNamespaceService.cs [hunk 1]
```diff
@@ -33,11 +33,11 @@ internal abstract class AbstractChangeNamespaceService : IChangeNamespaceService
 
         /// <summary>
         /// Try to get a new node to replace given node, which is a reference to a top-level type declared inside the 
-        /// namespce to be changed. If this reference is the right side of a qualified name, the new node returned would
+        /// namespace to be changed. If this reference is the right side of a qualified name, the new node returned would
         /// be the entire qualified name. Depends on whether <paramref name="newNamespaceParts"/> is provided, the name 
         /// in the new node might be qualified with this new namespace instead.
         /// </summary>
-        /// <param name="reference">A reference to a type declared inside the namespce to be changed, which is calculated
```
Comment (ryzngard) [inline src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractChangeNamespaceService.cs]: I guess at least I'm consistent with typos? 🤷‍♀️

File: src/Features/Core/Portable/MoveToNamespace/AbstractMoveToNamespaceService.cs [hunk 2]
```diff
@@ -252,24 +253,44 @@ private static async Task<ImmutableArray<ISymbol>> GetMemberSymbolsAsync(Documen
                 moveSpan,
                 MoveTypeOperationKind.MoveTypeNamespaceScope,
                 cancellationToken).ConfigureAwait(false);
-
             var modifiedDocument = modifiedSolution.GetDocument(document.Id);
-            var syntaxRoot = await modifiedDocument.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
 
+            // Since MoveTypeService doesn't handle linked files, we need to merge the diff ourselves,
```
Comment (ryzngard) [inline src/Features/Core/Portable/MoveToNamespace/AbstractMoveToNamespaceService.cs]: Should MoveTypeService handle linked files?

Comment (genlu) [reply src/Features/Core/Portable/MoveToNamespace/AbstractMoveToNamespaceService.cs]: yes. And in my opinion, all services should handle linked files (similar to how ChangeNameSpace service works) as their results will be used as the base for further code change. Preferably, we should come up a shared service to handle this for them, which requires a bigger change. I will file an issue to track that.

Comment (genlu) [reply src/Features/Core/Portable/MoveToNamespace/AbstractMoveToNamespaceService.cs]: #40040

**Comment by genlu** [comment]
@ryzngard Any more comments?

**Comment by genlu** [comment]
Ping @ryzngard

**Comment by DemetrisLeptos** [comment]
the problem persists after VS update and a repair just in case

Microsoft Visual Studio Community 2019
Version 16.9.0

System.Runtime.InteropServices.COMException : Files and folders cannot:
-be empty strings
- contain any of the following characters: / ? : & \ * " < > | # %
- contain Unicode control characters
- contain surrogate characters
- be system reserved names, including 'CON', 'AUX', 'PRN', 'COM1' or 'LPT2'
- be '.' or '..'

Please enter a valid name.
   at EnvDTE.ProjectItems.AddFolder(String Name,String Kind)
   at Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.Extensions.ProjectExtensions.CreateFolder(ProjectItems currentItems,String container)
   at Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.Extensions.ProjectExtensions.FindOrCreateFolder(Project project,IEnumerable`1 containers)
   at Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioWorkspaceImpl.AddDocumentToFolder(Project project,DocumentId documentId,IEnumerable`1 folders,String documentName,TextDocumentKind documentKind,SourceText initialText,String filePath)
   at Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioWorkspaceImpl.AddDocumentCore(DocumentInfo info,SourceText initialText,TextDocumentKind documentKind)
   at Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioWorkspaceImpl.ApplyDocumentAdded(DocumentInfo info,SourceText text)
   at Microsoft.CodeAnalysis.Workspace.ApplyProjectChanges(ProjectChanges projectChanges)
   at Microsoft.CodeAnalysis.Workspace.TryApplyChanges(Solution newSolution,IProgressTracker progressTracker)
   at Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioWorkspaceImpl.TryApplyChanges(Solution newSolution,IProgressTracker progressTracker)
   at Microsoft.CodeAnalysis.CodeActions.ApplyChangesOperation.TryApply(Workspace workspace,IProgressTracker progressTracker,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Editor.Implementation.CodeActions.CodeActionEditHandlerService.ProcessOperations(Workspace workspace,ImmutableArray`1 operations,IProgressTracker progressTracker,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Editor.Implementation.CodeActions.CodeActionEditHandlerService.Apply(Workspace workspace,Document fromDocument,ImmutableArray`1 operations,String title,IProgressTracker progressTracker,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedAction.InvokeWorker(Func`1 getFromDocument,IProgressTracker progressTracker,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedAction.<>c__DisplayClass22_0.<InvokeCore>b__0()
   at Microsoft.CodeAnalysis.Extensions.IExtensionManagerExtensions.PerformAction(IExtensionManager extensionManager,Object extension,Action action)
   at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)

**Pull request merged**